Source/Dafny/Verifier/Translator.ClassMembers.cs

```diff
@@ -244,21 +426,7 @@ public partial class Translator {
     ///         (h[o, alloc] ==> $IsAlloc(h[o, f], TT(TClassA_Inv_i(dtype(o)),..), h)) &&
     ///         $Is(h[o, f], TT(TClassA_Inv_i(dtype(o)),..), h);
     /// <summary>
-    private void AddAllocationAxiom(Boogie.Declaration fieldDeclaration, Field f, TopLevelDeclWithMembers c, bool is_array = false) {
-      Contract.Requires(c != null);
-      // IFF you're adding the array axioms, then the field should be null
-      Contract.Requires(is_array == (f == null));
-      Contract.Requires(sink != null && predef != null);
-
-      Bpl.Expr heightAntecedent = Bpl.Expr.True;
-      if (f is ConstantField) {
-        var cf = (ConstantField)f;
-        AddWellformednessCheck(cf);
-        if (InVerificationScope(cf)) {
-          var etran = new ExpressionTranslator(this, predef, f.tok);
-          heightAntecedent = Bpl.Expr.Lt(Bpl.Expr.Literal(cf.EnclosingModule.CallGraph.GetSCCRepresentativePredecessorCount(cf)), etran.FunctionContextHeight());
-        }
-      }
+    private void AddStaticConstFieldAllocationAxiom(Boogie.Declaration fieldDeclaration, Field f, TopLevelDeclWithMembers c, Expr heightAntecedent) {
```

Comment:
I have no idea what you did here, but the body of this method is very different compared to the body of the method with the same name before.

Reply:
The method `AddAllocationAxiom` is now on line 201

Reply:
I did see the refactoring. I was actually looking for same method names and the previous body of `AddStaticConstFieldAllocationAxiom`, before the move, is the following
```
      var oDotF = new Boogie.NAryExpr(c.tok, new Boogie.FunctionCall(GetReadonlyField(f)), tyexprs);
      var is_hf = MkIs(oDotF, f.Type); // $Is(h[o, f], ..)
      Boogie.Expr ax = bvsTypeAxiom.Count == 0 ? is_hf : BplForall(bvsTypeAxiom, BplTrigger(oDotF), is_hf);
      var isAxiom = new Boogie.Axiom(c.tok, BplImp(heightAntecedent, ax), $"{c}.{f}: Type axiom");
      AddOtherDefinition(fieldDeclaration, isAxiom);
      if (CommonHeapUse || (NonGhostsUseHeap && !f.IsGhost)) {
        Boogie.Expr h;
        var hVar = BplBoundVar("$h", predef.HeapType, out h);
        bvsAllocationAxiom.Add(hVar);
        var isGoodHeap = FunctionCall(c.tok, BuiltinFunction.IsGoodHeap, null, h);
        var isalloc_hf = MkIsAlloc(oDotF, f.Type, h); // $IsAlloc(h[o, f], ..)
        ax = BplForall(bvsAllocationAxiom, BplTrigger(isalloc_hf), BplImp(isGoodHeap, isalloc_hf));
        var isAllocAxiom = new Boogie.Axiom(c.tok, BplImp(heightAntecedent, ax), $"{c}.{f}: Allocation axiom");
        sink.AddTopLevelDeclaration(isAllocAxiom);
      }
```
whereas the body of the same method `AddStaticConstFieldAllocationAxiom` after the move, is the following:
```

      var bvsTypeAxiom = new List<Bpl.Variable>();
      var bvsAllocationAxiom = new List<Bpl.Variable>();

      // G
      var tyvars = MkTyParamBinders(GetTypeParams(c), out var tyexprs);
      bvsTypeAxiom.AddRange(tyvars);
      bvsAllocationAxiom.AddRange(tyvars);

      var oDotF = new Boogie.NAryExpr(c.tok, new Boogie.FunctionCall(GetReadonlyField(f)), tyexprs);
      var is_hf = MkIs(oDotF, f.Type); // $Is(h[o, f], ..)
      Boogie.Expr ax = bvsTypeAxiom.Count == 0 ? is_hf : BplForall(bvsTypeAxiom, BplTrigger(oDotF), is_hf);
      var isAxiom = new Boogie.Axiom(c.tok, BplImp(heightAntecedent, ax), $"{c}.{f}: Type axiom");
      AddOtherDefinition(fieldDeclaration, isAxiom);

      if (CommonHeapUse || (NonGhostsUseHeap && !f.IsGhost)) {
        Boogie.Expr h;
        var hVar = BplBoundVar("$h", predef.HeapType, out h);
        bvsAllocationAxiom.Add(hVar);
        var isGoodHeap = FunctionCall(c.tok, BuiltinFunction.IsGoodHeap, null, h);
        var isalloc_hf = MkIsAlloc(oDotF, f.Type, h); // $IsAlloc(h[o, f], ..)
        ax = BplForall(bvsAllocationAxiom, BplTrigger(isalloc_hf), BplImp(isGoodHeap, isalloc_hf));
        var isAllocAxiom = new Boogie.Axiom(c.tok, BplImp(heightAntecedent, ax), $"{c}.{f}: Allocation axiom");
        sink.AddTopLevelDeclaration(isAllocAxiom);
      }
```

First, can you please expand the comment `// G` ?

Second, can me explain why you now assign in local variables what was previously parameters? Usually, refactoring adds parameters, rarely removes some.

Reply:
> Second, can me explain why you now assign in local variables what was previously parameters? Usually, refactoring adds parameters, rarely removes some.

I took a big method and extracted multiple methods out of that, and some variable definitions that were cheap to compute and shared between the extracted methods, I copied into each of the methods instead of adding them as parameters, since I think the readability cost of defining them in the methods is lower than adding them as parameters.